lxd/device/cdi/id.go

```diff
@@ -0,0 +1,189 @@
+type DeviceType string
+
+const (
+	// SimpleGPU represents a single discrete GPU card.
+	SimpleGPU DeviceType = "gpu"
+	// All represents all the GPUs on a system.
+	All DeviceType = "all"
+	// MIG represents a single MIG compatible GPU card.
+	MIG DeviceType = "mig"
+	// IGPU represents a single iGPU device.
+	IGPU DeviceType = "igpu"
+)
+
+// ToDeviceType converts a string to a CDI device type.
+func ToDeviceType(deviceType string) (DeviceType, error) {
+	switch deviceType {
+	case string(SimpleGPU):
+		return SimpleGPU, nil
+	case string(All):
+		return All, nil
+	case string(MIG):
+		return MIG, nil
+	case string(IGPU):
+		return IGPU, nil
+	default:
+		return "", fmt.Errorf("invalid CDI device type (%q)", deviceType)
+	}
+}
+
+// ID represents a Container Device Interface (CDI) identifier.
+//
+// +------------+---------+------------+---------------------+
+// |   Vendor   | Product | DeviceType |       Locator       |
+// +---------------------------------------------------------+
+// | nvidia.com |   gpu   |    gpu     | [dev_idx]           |
+// |            |         |    all     | nil                 |
+// |            |         |    mig     | [dev_idx]:[mig_idx] |
+// |            |         |    igpu    | [dev_idx]           |
+// +------------+---------+------------+---------------------+
+//
+// Examples:
+//   - nvidia.com/gpu=gpu0
+type ID interface {
+	Vendor() Vendor
+	Product() Product
+	DeviceType() DeviceType
+	Locator() []string
+}
+
+// commonGPUID represents a common GPU CDI identifier.
+type commonGPUID struct {
+	vendor     Vendor
+	product    Product
+	deviceType DeviceType
+	deviceIdx  string
+}
+
+// Vendor returns the vendor of the GPU.
+func (id commonGPUID) Vendor() Vendor {
+	return id.vendor
+}
+
+// Product returns the product type.
+func (id commonGPUID) Product() Product {
+	return id.product
+}
+
+// DeviceType returns the device type.
+func (id commonGPUID) DeviceType() DeviceType {
+	return id.deviceType
+}
+
+// Locator returns the device location.
+func (id commonGPUID) Locator() []string {
+	if id.deviceType == All {
+		return nil
+	}
+
+	return []string{id.deviceIdx}
+}
+
+// migGPUID represents a MIG GPU CDI identifier.
+type migGPUID struct {
+	commonGPUID
+
+	migIdx string
+}
+
+// Locator returns a slice containing the device index
+// in the first position and the MIG index in the second position.
+func (id migGPUID) Locator() []string {
+	return []string{id.deviceIdx, id.migIdx}
+}
+
+// ToCDI converts a string identifier to a CDI ID (this is vendor agnostic). Examples:
+//
+//	"nvidia.com/gpu=gpu0" -> commonGPUID{vendor: "nvidia.com", product: "gpu", deviceName: "gpu", deviceIdx: 0}
+//	"nvidia.com/gpu=mig0:0" -> migGPUID{commonGPUID: commonGPUID{vendor: "nvidia.com", product: "gpu", deviceName: "mig", deviceIdx: 0}, migIdx: 0}
+//	"amd.com/gpu=all" -> commonGPUID{vendor: "amd.com", product: "gpu", deviceName: "all", all: true}
+func ToCDI(id string) (ID, error) {
```

Comment:
Would the CDI `parser` package be useful here: https://github.com/cncf-tags/container-device-interface/blob/main/pkg/parser/parser.go#L51

This should mean that you don't have to reimplement the logic for what is a valid CDI device name.

Reply:
fixed